checks/src/main/kotlin/com/faithlife/lint/ForEachFunctionDetector.kt

```diff
@@ -0,0 +1,58 @@
+package com.faithlife.lint
+
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Incident
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.intellij.psi.PsiMethod
+import org.jetbrains.uast.UCallExpression
+import org.jetbrains.uast.kotlin.KotlinUSafeQualifiedExpression
+
+class ForEachFunctionDetector : Detector(), SourceCodeScanner {
+    override fun getApplicableMethodNames(): List<String> = listOf("forEach", "forEachIndexed")
+
+    override fun visitMethodCall(context: JavaContext, node: UCallExpression, method: PsiMethod) {
+        if (node.uastParent is KotlinUSafeQualifiedExpression) return
```

Comment:
I'm erring on the permissive side here, but really I think we ought to not have this exception. Since the main value of permitting this use of the function is safe null handing with a long chain of collection operators, here are a couple alternatives:

Example legitimate use with this exception in place

```
nullableList?.filter { it % 2 == 0 }?.map { it.hashCode() }?.forEach {
}
```

could be replaced by

```
nullableList?.filter { it % 2 == 0 }?.map { it.hashCode() }?.let {
    for (code in it) {
    }
}
```

or better yet

```
val codes = nullableList?.filter { it % 2 == 0 }?.map { it.hashCode() }
for (code in codes ?: emptyList()) {
}
```

I'll leave in the exception since this rule is only informational (a suggestion), but if we ever want to take this rule more seriously it would be a good idea to reconsider, I think.